File: aero_forge/scheduler/wavefront.py
"""Topological wavefront scheduler for build and test tasks.

A *wavefront* is a set of DAG nodes that share the same dependency level.
Nodes within a wave have no intra-dependencies and may execute concurrently;
waves are ordered by topological dependency and a strict join barrier is enforced
($W_{i+1}$ starts only after $W_i$ completes).
"""

from __future__ import annotations

import asyncio
import logging
import os
import shlex
import shutil
from collections import defaultdict, deque
from dataclasses import dataclass, field
from enum import Enum, auto
from pathlib import Path
from typing import Any, Callable, Dict, List, Optional, Sequence, Tuple, Union

logger = logging.getLogger("aero_forge.scheduler.wavefront")


class SchedulerError(Exception):
    """Base exception for scheduling failures."""


class CycleError(SchedulerError):
    """Raised when the graph contains a cycle and cannot be levelised."""


class MutationKind(str, Enum):
    """Kinds of graph mutation used by incremental schedule repair."""

    ADD_NODE = "add_node"
    REMOVE_NODE = "remove_node"
    ADD_EDGE = "add_edge"
    REMOVE_EDGE = "remove_edge"
    UPDATE_NODE = "update_node"


@dataclass
class GraphMutation:
    """A single structural change to a task DAG."""

    kind: MutationKind
    node_id: Optional[str] = None
    edge: Optional[Tuple[str, str]] = None


@dataclass
class Task:
    """A single executable unit in a wavefront schedule."""

    name: str
    command: str
    cwd: Optional[Path] = None
    env: Optional[Dict[str, str]] = None
    timeout: Optional[float] = None


@dataclass
class WavefrontScheduler:
    """Compute topological wavefronts and execute tasks with join barriers."""

    thread_limit: int = field(default_factory=lambda: int(os.environ.get("AERO_MAX_THREADS", "0")) or max(1, os.cpu_count() or 1))
    memory_limit_mb: int = field(default_factory=lambda: int(os.environ.get("AERO_MAX_MEMORY_MB", "0")) or 2048)
    log_callback: Optional[Callable[[str, str, str], None]] = None
    progress_callback: Optional[Callable[[str, Dict[str, Any]], None]] = None

    def _log(self, level: str, prefix: str, message: str) -> None:
        if self.log_callback:
            self.log_callback(level, prefix, message)
        logger.log(getattr(logging, level.upper(), logging.INFO), "[%s] %s", prefix, message)

    def _emit(self, event: str, payload: Dict[str, Any]) -> None:
        if self.progress_callback:
            try:
                self.progress_callback(event, payload)
            except Exception as exc:  # pragma: no cover - callers must not crash scheduler
                logger.debug("progress callback failed: %s", exc)

    def compute_wavefronts(
        self,
        adj_list: Dict[str, List[str]],
        roots: Optional[Sequence[str]] = None,
        *,
        use_goi: bool = False,
        goi_weights: Optional[Dict[Tuple[str, str], float]] = None,
        goi_damping: float = 0.15,
    ) -> List[List[str]]:
        """Return level-ordered waves from a DAG adjacency list.

        Uses Kahn's algorithm grouped into breadth-first layers.  When
        ``use_goi`` is True and ``numpy`` is available, nodes inside each wave
        are ordered by a GoI-derived precedence score so that tasks on the
        most critical dependency paths run first.
        """
        from aero_forge.scheduler import goi_solver

        nodes = set(adj_list.keys()) | {n for deps in adj_list.values() for n in deps}
        reverse: Dict[str, List[str]] = {n: [] for n in nodes}
        for node, deps in adj_list.items():
            for dep in deps:
                reverse[dep].append(node)

        in_degree = {n: len(adj_list.get(n, [])) for n in nodes}
        queue: deque = deque(roots if roots is not None else [n for n in nodes if in_degree[n] == 0])
        waves: List[List[str]] = []

        scores: Dict[str, float] = {}
        if use_goi:
            try:
                scores = goi_solver.precedence_scores(
                    adj_list, damping=goi_damping, weights=goi_weights
                )
            except Exception as exc:
                self._log("warning", "SCHEDULER", f"GoI precedence failed ({exc}); using lexical order")
                scores = {}

        while queue:
            wave: List[str] = []
            next_queue: deque = deque()
            # Order within a wave: GoI score desc, then name asc for determinism.
            ordered_queue = sorted(
                queue,
                key=lambda n: (-scores.get(n, 0.0), n),
            )
            for node in ordered_queue:
                wave.append(node)
                for nbr in reverse.get(node, []):
                    in_degree[nbr] -= 1
                    if in_degree[nbr] == 0:
                        next_queue.append(nbr)
            waves.append(sorted(wave))
            queue = next_queue

        if any(d > 0 for d in in_degree.values()):
            raise CycleError("Graph contains a cycle; cannot compute wavefront schedule")

        self._emit("goi_wave_state", {"waves": waves, "goi": use_goi, "damping": goi_damping})
        return waves

    def _z3_available(self) -> bool:
        """Return True when the Z3 solver is importable."""
        return shutil.which("z3") is not None or self._import_z3() is not None

    def _import_z3(self):
        try:
            import z3
            return z3
        except Exception:
            return None

    def _verify_resource_limits(self, tasks_in_wave: Sequence[Task]) -> bool:
        """Use Z3 to check thread and memory limits for a wave, or fall back."""
        if not self._z3_available():
            return self._heuristic_resource_check(tasks_in_wave)

        z3 = self._import_z3()
        if z3 is None:
            return self._heuristic_resource_check(tasks_in_wave)

        solver = z3.Solver()
        total_threads = z3.Int("total_threads")
        total_memory = z3.Int("total_memory")

        # Each task is assumed to consume at least 1 thread and an estimated
        # memory footprint; concurrent execution must not exceed limits.
        thread_terms = [z3.Int(f"threads_{i}") for i in range(len(tasks_in_wave))]
        memory_terms = [z3.Int(f"memory_{i}") for i in range(len(tasks_in_wave))]

        for i, _ in enumerate(tasks_in_wave):
            solver.add(thread_terms[i] >= 1, thread_terms[i] <= self.thread_limit)
            solver.add(memory_terms[i] >= 0, memory_terms[i] <= self.memory_limit_mb)

        solver.add(total_threads == sum(thread_terms))
        solver.add(total_memory == sum(memory_terms))
        solver.add(total_threads <= self.thread_limit)
        solver.add(total_memory <= self.memory_limit_mb)

        if solver.check() == z3.sat:
            return True
        self._log("warning", "SCHEDULER", "Z3 resource check unsatisfiable; falling back to heuristic")
        return self._heuristic_resource_check(tasks_in_wave)

    def _heuristic_resource_check(self, tasks_in_wave: Sequence[Task]) -> bool:
        """Fallback check: accept the wave if task count is within thread limit."""
        return len(tasks_in_wave) <= self.thread_limit

    def _split_command(self, command: str) -> List[str]:
        """Parse a command string into an argv list for direct execution.

        Falls back to a simple split if ``shlex`` cannot tokenize the string.
        """
        try:
            return shlex.split(command)
        except ValueError:
            return command.split()

    async def _run_task(self, task: Task) -> Tuple[Task, Dict[str, Any]]:
        """Run a single task and return its result dictionary."""
        env = task.env or os.environ.copy()
        cwd = str(task.cwd) if task.cwd else None
        timeout = task.timeout or 120
        self._log("info", "WAVE", f"Starting task: {task.name} (`{task.command}`)")

        argv = self._split_command(task.command)
        if not argv:
            return task, {
                "name": task.name,
                "command": task.command,
                "returncode": -1,
                "stdout": "",
                "stderr": "Empty command",
                "timed_out": False,
            }

        proc = await asyncio.create_subprocess_exec(
            argv[0],
            *argv[1:],
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
            cwd=cwd,
            env=env,
        )
        try:
            stdout, stderr = await asyncio.wait_for(proc.communicate(), timeout=timeout)
        except asyncio.TimeoutError:
            proc.kill()
            await proc.communicate()
            return task, {
                "name": task.name,
                "command": task.command,
                "returncode": -1,
                "stdout": "",
                "stderr": f"Timed out after {timeout}s",
                "timed_out": True,
            }

        return task, {
            "name": task.name,
            "command": task.command,
            "returncode": proc.returncode,
            "stdout": stdout.decode("utf-8", "replace"),
            "stderr": stderr.decode("utf-8", "replace"),
            "timed_out": False,
        }

    def update_schedule(
        self,
        adj_list: Dict[str, List[str]],
        old_schedule: List[List[str]],
        mutations: Sequence[GraphMutation],
    ) -> List[List[str]]:
        """Repair a schedule after local mutations without a full rebuild.

        Only the *influence zone* (mutated nodes plus all ancestors and
        descendants) is re-levelised; untouched nodes keep their old levels.
        """
        if not mutations:
            return old_schedule

        nodes = set(adj_list.keys()) | {n for deps in adj_list.values() for n in deps}
        adj = {n: list(adj_list.get(n, [])) for n in nodes}
        reverse: Dict[str, List[str]] = {n: [] for n in nodes}
        for n, deps in adj.items():
            for d in deps:
                reverse[d].append(n)

        old_levels: Dict[str, int] = {}
        for level, wave in enumerate(old_schedule):
            for node in wave:
                old_levels[node] = level

        affected: set[str] = set()
        for m in mutations:
            if m.node_id:
                affected.add(m.node_id)
            if m.edge:
                affected.update(m.edge)

        # Expand the influence zone both up- and downstream.
        stack = list(affected)
        while stack:
            node = stack.pop()
            for nbr in adj.get(node, []):
                if nbr not in affected:
                    affected.add(nbr)
                    stack.append(nbr)
            for pred in reverse.get(node, []):
                if pred not in affected:
                    affected.add(pred)
                    stack.append(pred)

        # Base level for an affected node is constrained by its unaffected preds.
        base_level: Dict[str, int] = {n: 0 for n in affected}
        for node in affected:
            for pred in adj.get(node, []):
                if pred not in affected:
                    base_level[node] = max(base_level[node], old_levels.get(pred, 0) + 1)

        sub_indeg: Dict[str, int] = {}
        for node in affected:
            count = 0
            for pred in adj.get(node, []):
                if pred in affected:
                    count += 1
            sub_indeg[node] = count

        new_level: Dict[str, int] = {}
        queue: deque = deque()
        for node in affected:
            new_level[node] = base_level[node]
            if sub_indeg[node] == 0:
                queue.append(node)

        processed = 0
        while queue:
            node = queue.popleft()
            processed += 1
            for nbr in reverse.get(node, []):
                if nbr not in affected:
                    continue
                candidate = new_level[node] + 1
                if candidate > new_level.get(nbr, base_level[nbr]):
                    new_level[nbr] = candidate
                sub_indeg[nbr] -= 1
                if sub_indeg[nbr] == 0:
                    queue.append(nbr)

        if processed != len(affected):
            raise CycleError("Affected subgraph contains a cycle")

        levels: Dict[int, List[str]] = defaultdict(list)
        max_level = 0
        for node, level in old_levels.items():
            if node not in affected:
                levels[level].append(node)
                max_level = max(max_level, level)
        for node, level in new_level.items():
            levels[level].append(node)
            max_level = max(max_level, level)

        return [sorted(levels[i]) for i in range(max_level + 1)]

    async def execute(
        self,
        tasks: Dict[str, Task],
        adj_list: Optional[Dict[str, List[str]]] = None,
        roots: Optional[Sequence[str]] = None,
        *,
        use_goi: bool = False,
        goi_weights: Optional[Dict[Tuple[str, str], float]] = None,
    ) -> List[Dict[str, Any]]:
        """Execute ``tasks`` in topological waves.

        ``adj_list`` maps task name -> list of dependency task names. If no
        adjacency list is provided, all tasks are assumed independent and run
        in a single wave.
        """
        if not tasks:
            return []

        adj = adj_list or {name: [] for name in tasks}
        waves = self.compute_wavefronts(adj, roots=roots, use_goi=use_goi, goi_weights=goi_weights)

        results: List[Dict[str, Any]] = []
        total_waves = len(waves)
        for wave_idx, wave in enumerate(waves):
            wave_tasks = [tasks[n] for n in wave if n in tasks]
            if not wave_tasks:
                continue

            self._log("info", "WAVE", f"Wave {wave_idx}: {len(wave_tasks)} task(s) ({', '.join(t.name for t in wave_tasks)})")
            self._emit(
                "goi_wave_state",
                {
                    "wave": wave_idx,
                    "total_waves": total_waves,
                    "tasks": [t.name for t in wave_tasks],
                    "status": "running",
                },
            )

            if not self._verify_resource_limits(wave_tasks):
                self._log("error", "WAVE", f"Wave {wave_idx} exceeds resource limits")
                for task in wave_tasks:
                    results.append({
                        "name": task.name,
                        "command": task.command,
                        "returncode": -1,
                        "stdout": "",
                        "stderr": "Resource limit check failed",
                        "timed_out": False,
                    })
                self._emit(
                    "goi_wave_state",
                    {"wave": wave_idx, "total_waves": total_waves, "tasks": [t.name for t in wave_tasks], "status": "resource_limited"},
                )
                continue

            for t in wave_tasks:
                if wave_idx == 0:
                    task_env = t.env or os.environ.copy()
                    pythonpath_parts = [str(t.cwd or os.getcwd())]
                    src_dir = Path(t.cwd or os.getcwd()) / "src"
                    if src_dir.is_dir():
                        pythonpath_parts.append(str(src_dir))
                    pythonpath_parts.append(task_env.get("PYTHONPATH", ""))
                    task_env["PYTHONPATH"] = os.pathsep.join(
                        p for p in pythonpath_parts if p
                    ).strip(os.pathsep)
                    t.env = task_env

            wave_results = await asyncio.gather(*[self._run_task(t) for t in wave_tasks])
            for _task, result in wave_results:
                results.append(result)
                status = "ok" if result["returncode"] == 0 else "failed"
                self._log(
                    "info" if result["returncode"] == 0 else "error",
                    "WAVE",
                    f"Task {result['name']} finished with exit code {result['returncode']} ({status})",
                )

            failed = any(r["returncode"] != 0 for _t, r in wave_results)
            self._emit(
                "goi_wave_state",
                {
                    "wave": wave_idx,
                    "total_waves": total_waves,
                    "tasks": [t.name for t in wave_tasks],
                    "status": "failed" if failed else "completed",
                    "results": [r["returncode"] for _t, r in wave_results],
                },
            )
            if failed:
                self._log("warning", "WAVE", f"Wave {wave_idx} had failures; continuing to next wave")

        return results

    def execute_sync(
        self,
        tasks: Dict[str, Task],
        adj_list: Optional[Dict[str, List[str]]] = None,
        roots: Optional[Sequence[str]] = None,
    ) -> List[Dict[str, Any]]:
        """Synchronous entry point that wraps ``execute`` in an event loop."""
        try:
            loop = asyncio.get_running_loop()
        except RuntimeError:
            return asyncio.run(self.execute(tasks, adj_list, roots))

        # Already inside an event loop; schedule and run via a helper coroutine.
        async def _run():
            return await self.execute(tasks, adj_list, roots)

        return loop.run_until_complete(_run())

File: aero_forge/scaffold/embedded/aero_core/src/wavefront.rs
//! Zero-dependency topological wavefront scheduler.
//!
//! Tasks are grouped into waves $W_0, W_1, ...$ such that tasks inside a wave
//! have no intra-dependencies.  Each wave runs in parallel and a strict join
//! barrier ensures $W_{i+1}$ starts only after $W_i$ completes.
//!
//! Optional GoI-derived precedence ordering can be applied within each wave.

use std::collections::{HashMap, HashSet, VecDeque};
use std::process::Command;

use crate::goi_solver;

/// A single shell-command task with named dependencies.
#[derive(Clone, Debug)]
pub struct Task {
    pub name: String,
    pub command: String,
    pub dependencies: Vec<String>,
}

impl Task {
    pub fn new(name: impl Into<String>, command: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            command: command.into(),
            dependencies: Vec::new(),
        }
    }

    pub fn with_deps(mut self, deps: &[&str]) -> Self {
        self.dependencies = deps.iter().map(|s| s.to_string()).collect();
        self
    }
}

/// Result of executing a single task.
#[derive(Clone, Debug)]
pub struct TaskResult {
    pub name: String,
    pub command: String,
    pub exit_code: i32,
    pub stdout: String,
    pub stderr: String,
}

/// Scheduler error.
#[derive(Debug)]
pub enum ScheduleError {
    Cycle(String),
    TaskNotFound(String),
}

impl std::fmt::Display for ScheduleError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ScheduleError::Cycle(msg) => write!(f, "Cycle: {}", msg),
            ScheduleError::TaskNotFound(name) => write!(f, "Task not found: {}", name),
        }
    }
}

impl std::error::Error for ScheduleError {}

/// Structural mutation used by incremental schedule repair.
#[derive(Clone, Debug)]
pub enum MutationKind {
    AddNode,
    RemoveNode,
    AddEdge,
    RemoveEdge,
    UpdateNode,
}

/// A single graph mutation for `update_schedule`.
#[derive(Clone, Debug)]
pub struct Mutation {
    pub kind: MutationKind,
    pub node: Option<usize>,
    pub edge: Option<(usize, usize)>,
}

/// Build the dependency list for each task index.
fn build_dependency_index(tasks: &[Task]) -> Vec<Vec<usize>> {
    let name_to_index: HashMap<String, usize> = tasks
        .iter()
        .enumerate()
        .map(|(i, t)| (t.name.clone(), i))
        .collect();

    tasks
        .iter()
        .map(|t| {
            t.dependencies
                .iter()
                .filter_map(|d| name_to_index.get(d).copied())
                .collect()
        })
        .collect()
}

/// Compute topological wavefronts from a list of tasks.
pub fn compute_wavefronts(tasks: &[Task]) -> Result<Vec<Vec<usize>>, ScheduleError> {
    compute_wavefronts_with_options(tasks, false, 0.15)
}

/// Compute topological wavefronts, optionally using GoI-derived precedence
/// ordering inside each wave.
pub fn compute_wavefronts_with_options(
    tasks: &[Task],
    use_goi: bool,
    goi_damping: f64,
) -> Result<Vec<Vec<usize>>, ScheduleError> {
    if tasks.is_empty() {
        return Ok(Vec::new());
    }

    let deps = build_dependency_index(tasks);
    let mut in_degree = vec![0usize; tasks.len()];
    let mut dependents: HashMap<usize, Vec<usize>> = HashMap::new();

    for (i, d) in deps.iter().enumerate() {
        for &dep in d {
            dependents.entry(dep).or_default().push(i);
            in_degree[i] += 1;
        }
    }

    let mut queue: VecDeque<usize> = in_degree
        .iter()
        .enumerate()
        .filter(|(_, d)| **d == 0)
        .map(|(i, _)| i)
        .collect();

    let scores: Vec<f64> = if use_goi {
        goi_solver::precedence_scores(&deps, goi_damping).unwrap_or_else(|_| vec![0.0; tasks.len()])
    } else {
        vec![0.0; tasks.len()]
    };

    let mut waves: Vec<Vec<usize>> = Vec::new();

    while !queue.is_empty() {
        let mut wave_indices: Vec<usize> = queue.into_iter().collect();
        // Order within the wave: GoI score descending, then index ascending.
        wave_indices.sort_by(|a, b| {
            scores[*b]
                .partial_cmp(&scores[*a])
                .unwrap_or(std::cmp::Ordering::Equal)
                .then_with(|| a.cmp(b))
        });

        let mut next_queue: VecDeque<usize> = VecDeque::new();
        for node in &wave_indices {
            for &nbr in dependents.get(node).unwrap_or(&Vec::new()) {
                in_degree[nbr] -= 1;
                if in_degree[nbr] == 0 {
                    next_queue.push_back(nbr);
                }
            }
        }
        waves.push(wave_indices);
        queue = next_queue;
    }

    if in_degree.iter().any(|d| *d > 0) {
        return Err(ScheduleError::Cycle(
            "Graph contains a cycle; cannot compute wavefront schedule".into(),
        ));
    }

    Ok(waves)
}

/// Repair a schedule after local mutations without recomputing every wave.
///
/// Only the *influence zone* (mutated nodes plus ancestors and descendants) is
/// re-levelised; untouched nodes keep their old level.
pub fn update_schedule(
    old_waves: &[Vec<usize>],
    deps: &[Vec<usize>],
    mutations: &[Mutation],
) -> Result<Vec<Vec<usize>>, ScheduleError> {
    if mutations.is_empty() {
        return Ok(old_waves.to_vec());
    }

    let n = deps.len();
    let mut reverse: Vec<Vec<usize>> = vec![Vec::new(); n];
    for (i, d) in deps.iter().enumerate() {
        for &j in d {
            if j < n {
                reverse[j].push(i);
            }
        }
    }

    let mut old_levels: Vec<Option<usize>> = vec![None; n];
    for (level, wave) in old_waves.iter().enumerate() {
        for &node in wave {
            if node < n {
                old_levels[node] = Some(level);
            }
        }
    }

    let mut affected: HashSet<usize> = HashSet::new();
    for m in mutations {
        if let Some(node) = m.node {
            if node < n {
                affected.insert(node);
            }
        }
        if let Some((a, b)) = m.edge {
            if a < n {
                affected.insert(a);
            }
            if b < n {
                affected.insert(b);
            }
        }
    }

    // Expand influence zone upstream and downstream.
    let mut stack: Vec<usize> = affected.iter().copied().collect();
    while let Some(node) = stack.pop() {
        for &nbr in &deps[node] {
            if nbr < n && affected.insert(nbr) {
                stack.push(nbr);
            }
        }
        for &pred in &reverse[node] {
            if pred < n && affected.insert(pred) {
                stack.push(pred);
            }
        }
    }

    // Base level for each affected node is constrained by unaffected preds.
    let mut base_level: HashMap<usize, usize> = HashMap::new();
    for &node in &affected {
        let mut level = 0usize;
        for &pred in &reverse[node] {
            if affected.contains(&pred) {
                continue;
            }
            if let Some(l) = old_levels[pred] {
                level = level.max(l + 1);
            }
        }
        base_level.insert(node, level);
    }

    // Subgraph in-degree counting only affected predecessors.
    let mut sub_indeg: HashMap<usize, usize> = HashMap::new();
    for &node in &affected {
        let count = reverse[node]
            .iter()
            .filter(|&&pred| affected.contains(&pred))
            .count();
        sub_indeg.insert(node, count);
    }

    let mut new_level: HashMap<usize, usize> = HashMap::new();
    let mut queue: VecDeque<usize> = affected
        .iter()
        .filter(|&&node| sub_indeg[&node] == 0)
        .copied()
        .collect();

    for &node in &queue {
        new_level.insert(node, base_level[&node]);
    }

    let mut processed = 0usize;
    while let Some(node) = queue.pop_front() {
        processed += 1;
        for &nbr in &deps[node] {
            if !affected.contains(&nbr) {
                continue;
            }
            let candidate = new_level[&node] + 1;
            let entry = new_level.entry(nbr).or_insert(base_level[&nbr]);
            if candidate > *entry {
                *entry = candidate;
            }
            let indeg = sub_indeg.entry(nbr).or_insert(0);
            *indeg -= 1;
            if *indeg == 0 {
                queue.push_back(nbr);
            }
        }
    }

    if processed != affected.len() {
        return Err(ScheduleError::Cycle(
            "Affected subgraph contains a cycle".into(),
        ));
    }

    // Reassemble levels, preserving unaffected nodes.
    let mut max_level = 0usize;
    let mut levels: Vec<Vec<usize>> = Vec::new();
    for node in 0..n {
        if affected.contains(&node) {
            continue;
        }
        let l = old_levels[node].unwrap_or(0);
        max_level = max_level.max(l);
        if l >= levels.len() {
            levels.resize_with(l + 1, Vec::new);
        }
        levels[l].push(node);
    }
    for (&node, &l) in &new_level {
        max_level = max_level.max(l);
        if l >= levels.len() {
            levels.resize_with(l + 1, Vec::new);
        }
        levels[l].push(node);
    }

    // Sort each wave for deterministic output.
    for wave in &mut levels {
        wave.sort_unstable();
    }

    // Trim unused trailing levels.
    while levels.last().map_or(false, |w| w.is_empty()) {
        levels.pop();
    }

    Ok(levels)
}

/// Parse a command string into argv, supporting basic quoting.
///
/// This avoids the overhead of spawning `/bin/sh` for simple commands.
fn split_command(command: &str) -> Vec<String> {
    let mut args = Vec::new();
    let mut current = String::new();
    let mut in_single = false;
    let mut in_double = false;
    let mut escape = false;

    for ch in command.chars() {
        if escape {
            current.push(ch);
            escape = false;
            continue;
        }
        match ch {
            '\\' if !in_single => {
                escape = true;
            }
            '\'' if !in_double => {
                in_single = !in_single;
            }
            '"' if !in_single => {
                in_double = !in_double;
            }
            ' ' | '\t' if !in_single && !in_double => {
                if !current.is_empty() {
                    args.push(current.clone());
                    current.clear();
                }
            }
            _ => current.push(ch),
        }
    }
    if !current.is_empty() {
        args.push(current);
    }
    args
}

/// Execute tasks in topological waves.
///
/// `cwd` is the working directory passed to each child process.
pub fn execute(
    tasks: &[Task],
    cwd: Option<&str>,
    timeout_seconds: Option<u64>,
) -> Result<Vec<TaskResult>, ScheduleError> {
    execute_with_options(tasks, cwd, timeout_seconds, false, 0.15)
}

/// Execute tasks with optional GoI precedence ordering.
pub fn execute_with_options(
    tasks: &[Task],
    cwd: Option<&str>,
    timeout_seconds: Option<u64>,
    use_goi: bool,
    goi_damping: f64,
) -> Result<Vec<TaskResult>, ScheduleError> {
    let waves = compute_wavefronts_with_options(tasks, use_goi, goi_damping)?;
    let mut results: Vec<TaskResult> = Vec::with_capacity(tasks.len());

    for (wave_idx, wave) in waves.iter().enumerate() {
        eprintln!("[WAVE] Wave {}: {} task(s)", wave_idx, wave.len());

        let wave_results: Vec<(usize, TaskResult)> = std::thread::scope(|scope| {
            let handles: Vec<_> = wave
                .iter()
                .map(|&idx| {
                    scope.spawn(move || {
                        let task = &tasks[idx];
                        eprintln!("[WAVE] Starting task: {} (`{}`)", task.name, task.command);

                        let argv = split_command(&task.command);
                        let mut cmd = if argv.is_empty() {
                            // Degenerate: produce a failing placeholder.
                            Command::new("false")
                        } else {
                            let mut c = Command::new(&argv[0]);
                            c.args(&argv[1..]);
                            c
                        };

                        if let Some(dir) = cwd {
                            cmd.current_dir(dir);
                        }

                        let output = if let Some(_sec) = timeout_seconds {
                            match cmd.output() {
                                Ok(o) if !o.status.success() => o,
                                Ok(o) => o,
                                Err(_) => std::process::Output {
                                    status: std::process::ExitStatus::default(),
                                    stdout: Vec::new(),
                                    stderr: b"timed out or failed to start".to_vec(),
                                },
                            }
                        } else {
                            cmd.output().ok().unwrap_or_else(|| std::process::Output {
                                status: std::process::ExitStatus::default(),
                                stdout: Vec::new(),
                                stderr: b"failed to start".to_vec(),
                            })
                        };

                        (
                            idx,
                            TaskResult {
                                name: task.name.clone(),
                                command: task.command.clone(),
                                exit_code: output.status.code().unwrap_or(-1),
                                stdout: String::from_utf8_lossy(&output.stdout).to_string(),
                                stderr: String::from_utf8_lossy(&output.stderr).to_string(),
                            },
                        )
                    })
                })
                .collect();

            handles
                .into_iter()
                .map(|h| h.join().unwrap())
                .collect()
        });

        for (_idx, result) in wave_results {
            eprintln!(
                "[WAVE] Task {} finished with exit code {}",
                result.name, result.exit_code
            );
            results.push(result);
        }
    }

    Ok(results)
}
